netaddr.go

```diff
@@ -846,3 +871,422 @@ func (p *IPPrefix) UnmarshalText(text []byte) error {
 func (p IPPrefix) String() string {
 	return fmt.Sprintf("%s/%d", p.IP, p.Bits)
 }
+
+// LastIP returns the last IP in the prefix.
+func (p IPPrefix) LastIP() IP {
+	if p.IP.IsZero() {
+		return IP{}
+	}
+	a16 := p.IP.As16()
+	var off uint8
+	var bits uint8 = 128
+	if p.IP.Is4() {
+		off = 12
+		bits = 32
+	}
+	for b := p.Bits; b < bits; b++ {
+		byteNum, bitInByte := b/8, 7-(b%8)
+		a16[off+byteNum] |= 1 << uint(bitInByte)
+	}
+	if p.IP.Is4() {
+		return IPFrom16(a16)
+	} else {
+		return IPv6Raw(a16) // doesn't unmap
+	}
+}
+
+// IPRange represents an inclusive range of IP addresses
+// from the same address family.
+type IPRange struct {
+	// From is the starting IP address in the range.
+	// It must have the same address family (IPv4 vs IPv6) as To
+	// and be less than or equal to To.
+	// From is an inclusive bound; it is included in the set.
+	From IP
+
+	// To is the ending IP address in the range.
+	// It must have the same address family (IPv4 vs IPv6) as From
+	// and be greater than or equal to From.
+	// To is an inclusive bound; it is included in the set.
+	To IP
+}
+
+// Valid reports whether r.From and r.To are both non-zero and obey
+// the documented requirements: address families match, and From is
+// less than or equal to To.
+func (r IPRange) Valid() bool {
```

Comment:
Good call, I was going to ask why From/To were exported if the less-than invariant must hold. Can you think of any use case in particular where someone would construct an IPRange manually without a constructor function? If not, it seems tempting to me to unexport the fields.

Reply:
Well, we have IPPort already that's also with two exported fields. Felt consistent?

Reply:
Also IPPrefix, which can have invalid values (bits over 128) from literal construction.